src/eventQueue.ts

```diff
@@ -9,65 +9,44 @@ dotenv.config();
 
 class EventQueue extends EventEmitter {
   queue: ContractEvent[];
+  processing: boolean;
 
   constructor(options?: Record<string, unknown>) {
     super(options);
     this.queue = [];
+    this.processing = false;
   }
-}
-
-const eventQueue = new EventQueue();
 
-/*
- * Process a individual queue event
- */
-eventQueue.on(
-  QueueEvents.ProcessEvent,
-  function (this: EventQueue, event: ContractEvent) {
-    verbose('Processing event:', event.signature);
-    eventProcessor(event);
-  },
-);
-
-/*
- * @TODO Add batching logic (if needed)
- *
- * Start processing events from the queue
- */
-eventQueue.on(QueueEvents.ProcessEvents, function (this: EventQueue) {
-  verbose('Processing', this.queue.length, 'events in queue');
-  for (let index = 0; index < this.queue.length; index += 1) {
-    this.emit(QueueEvents.ProcessEvent, this.queue.shift());
+  async processEvents(): Promise<void> {
+    this.processing = true;
+    while (this.queue.length) {
```

Comment:
Because the reference to `this.queue` is dynamic, it means this loop will respond to all calls to `addEvent` for as long as it is running. This means we only have to call `processEvents` once, and it will process all events in the queue for as long as there are events in the queue, even if events are added after `processEvents` was initially called.